### DISCUSSION TITLE
Facing errors/issues while using skforecast

### DISCUSSION THREAD

**Discussion by ABJ66** [JoaquinAmatRodrigo/skforecast#569] [Q&A] [answered]
Hi all, 

I am new to skforecast and would like to learn & implement some of the functionalities in this package. 

Data:  7 products sales spanning from 2019-01 to 2023-03 at monthly level. These products are related to the same category/domain.

Ask: 1. Forecast individual product sales (multi-step) for next 12 months or so. 2. Later create forecasts at place of sales/stores level.

Techniques tried: Dependent Multi-Series Forecasting with one key product set to 'level', say 'ABC'

Questions: I've been going through some documentation (https://skforecast.org/0.10.1/user_guides/dependent-multi-series-multivariate-forecasting.html) and trying to test it which has given me some errors/issues. I am using Python: 3.9.5 and skforecast 0.10.1 on Databricks notebook

1. The predictions generated using forecaster.predict() do not return a time index (but are serially ordered integers), even though the train and test data have the same, with a name: 'date'. Is there a way to get time indicies?
2. While using this line: from skforecast.model_selection_multiseries import backtesting_forecaster_multiseries - I get the following error: dataclass_transform() got an unexpected keyword argument 'field_specifiers'
3. Also, does one need to change the level from 'ABC' to another one in the list of product to get predictions for that product and likewise for all such products?
4. Is there a way to perform hierarchical forecasting (top down) to get sales at store level, part of skforecast?

Thank you for your patience and support on these questions.

Regards!

**Comment by JoaquinAmatRodrigo** [comment]
Hi @ABJ66,
Thanks for using skforecast.

**The predictions generated using forecaster.predict() do not return a time index (but are serially ordered integers), even though the train and test data have the same, with a name: 'date'. Is there a way to get time indicies?**

This may be because the time series index has no frequency. You can set it using the `asfreq()` method. You can check the current frequency with `data.index.freq`.

**While using this line: from skforecast.model_selection_multiseries import backtesting_forecaster_multiseries - I get the following error: dataclass_transform() got an unexpected keyword argument 'field_specifiers'**

Could you please provide a reproducible example and the error you are getting? Without this, it is hard to know what the problem might be.

**Also, does one need to change the level from 'ABC' to another one in the list of product to get predictions for that product and likewise for all such products?**

The current implementation of `ForecasterAutoregMultiVariate` is designed to model a single time series using multiple time series. So the model actually learns to predict the time series specified in the `level` argument. If you want to predict multiple products, you need to train multiple models (each one to predict one product, but all of them using all the series to learn).

**Is there a way to perform hierarchical forecasting (top down) to get sales at store level, part of skforecast?**

Unfortunately, we do not have this feature in skforecast yet, but it would be great to have it. Do you have any experience with it? If you have any recommendations, we would be happy to review them.

Hope these answers help,

**Reply by ABJ66** [reply]
Hi Joaquin, 

Thank you for your response. Pls find below information which might be helpful to resolve these queries:

**1. The predictions generated using forecaster.predict() do not return a time index (but are serially ordered integers), even though the train and test data have the same, with a name: 'date'. Is there a way to get time indicies?**

I am able to get month*year combination as seen by printing the index for the test set: PeriodIndex(['2022-12', '2023-01', '2023-02', '2023-03'], dtype='period[M]', name='Date') and also with the output of the print statements: 

Train dates : 2019-01 --- 2022-12(n=48)
Test dates  : 2023-01 --- 2023-03(n=3)

However, I still get the integer based indicies for the predictions. Not sure what is still being missed out.

**2. For importing the submodules for backtesting, I am sharing the snippet of error:** 

TypeError: dataclass_transform() got an unexpected keyword argument 'field_specifiers'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File <command-1113636422469495>:1
----> 1 from skforecast.model_selection_multiseries import backtesting_forecaster_multiseries

File /databricks/python_shell/dbruntime/PythonPackageImportsInstrumentation/__init__.py:171, in _create_import_patch.<locals>.import_patch(name, globals, locals, fromlist, level)
    166 thread_local._nest_level += 1
    168 try:
    169     # Import the desired module. If you’re seeing this while debugging a failed import,
    170     # look at preceding stack frames for relevant error information.
--> 171     original_result = python_builtin_import(name, globals, locals, fromlist, level)
    173     is_root_import = thread_local._nest_level == 1
    174     # `level` represents the number of leading dots in a relative import statement.
    175     # If it's zero, then this is an absolute import.

Pls do let me know if there is a way to resolve the error.

Thanks in advance.

**Reply by JoaquinAmatRodrigo** [reply]
Hi @ABJ66 

Regarding the frequency. What is the output when you extract the freq of your data (`data.index.freq`)?

Regarding the import error, I do not have an answer. It is the first time we have seen this error. Have you tried to re-install the library?

**Reply by ABJ66** [reply]
Hello Joaquin: The frequency is "Month End". I have month*year numbers across sample products. Is it ok to work with?

**Reply by JoaquinAmatRodrigo** [reply]
Is "Month End" the output you get from (data.index.freq)?

**Reply by ABJ66** [reply]
Yes, Joaquin

**Reply by JavierEscobarOrtiz** [reply]
Hello @ABJ66,

The easiest way to solve this is converting to a pandas datetime object:

```python
# Data preprocessing
# ==============================================================================
data['datetime'] = pd.to_datetime(data['datetime'], format='%Y-%m-%d')
data = data.set_index('datetime')
data = data.asfreq('MS')
data = data.sort_index()
```

After it, your dates will look like this:

Train dates : 2019-01-01 --- 2022-12-01  (n=48)
Test dates : 2023-01-01 --- 2023-03-01  (n=3)

But at the end, the frequency is monthly (`MS`).

Just a tip, include '23:59:59'  when filtering the dates to avoid overlaps. Just like this example:

https://skforecast.org/latest/user_guides/backtesting#data

Hope it helps!

**Reply by JoaquinAmatRodrigo** [reply]
Thanks for checking.
There may be a bug in the code. Could you share a reproducible example? This way we can debug it.

**Reply by ABJ66** [reply]
Hi Joaquin, 

Sharing sample records for your reference (calling it as df for further code/instructions given below):
	    PRD_1, PRD_2, PRD_3
Date
2019-01 2971, 1383, 4473
2019-02 2714, 1235, 4189
2019-03 2837, 1301, 4218
2019-04 2977, 1397, 4282
2019-05 2919, 1478, 4403
2019-06 2536, 1316, 4040
2019-07 2803, 1345, 4176
2019-08 2968, 1475, 4207
2019-09 2692, 1333, 4150
2019-10 2847, 1541, 4423
2019-11 2540, 1293, 3778
2019-12 2397, 1307, 3743
2020-01 2918, 1432, 4098
2020-02 2525, 1363, 3854
2020-03 2305, 1312, 3461
2020-04 1515, 918, 2453
2020-05 2297, 1219, 3497
2020-06 2813, 1432, 4113
2020-07 2736, 1413, 3828
2020-08 2551, 1286, 3515
2020-09 2541, 1358, 3749
2020-10 2834, 1552, 3883
2020-11 2260, 1243, 3392
2020-12 2366, 1320, 3369
2021-01 2275, 1122, 3621
...

end_train = '2019-12'
start_test = '2020-01'
data_train = df.loc[:end_train, :].copy()
data_test  = df.loc[start_test:, :].copy()

print(data_test.index.freq)

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
import skforecast

from skforecast.ForecasterAutoregMultiVariate import ForecasterAutoregMultiVariate
from skforecast.model_selection_multiseries import backtesting_forecaster_multiseries # this line throws an error as stated above

forecaster = ForecasterAutoregMultiVariate(
                 regressor          = Ridge(random_state=123),
                 level              = 'PRD_1',
                 lags               = 12,
                 steps              = 6,
                 transformer_series = None,
                 transformer_exog   = None,
                 weight_func        = None,
                 n_jobs             = 'auto'
             )

forecaster.fit(series=data_train)
predictions = forecaster.predict()
print(predictions)

The index for predictions dataframe/series shows integer based values rather than dates (Y-m) as stated in the original question.

Hope this helps. Thank you!

**Reply by ABJ66** [reply]
Hi again,

I tried to run the code locally and I see a warning message as below, seems the package only accepts either of the two options given below: This seems to be in function: **preprocess_y** (may be an additional inclusion in 'isinstance' of PeriodIndex could help?)

site-packages\skforecast\utils\utils.py:798: UserWarning: `y` has no **DatetimeIndex** nor **RangeIndex** index. Index is overwritten with a RangeIndex.
  warnings.warn(

**Reply by JoaquinAmatRodrigo** [reply]
Hi @ABJ66,
According to the example you share, your dataframe index does not have a frequency. Frequency is an attribute of the index that needs to be set by the user. As you can see in the following code, once it is established, the predictions have a datime index.

```
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
import skforecast
from skforecast.ForecasterAutoregMultiVariate import ForecasterAutoregMultiVariate
from skforecast.model_selection_multiseries import backtesting_forecaster_multiseries

data = {
        'PRD_1': [2971, 2714, 2837, 2977, 2919, 2536, 2803, 2968, 2692, 2847, 2540, 2397, 2918, 2525, 2305, 1515, 2297, 2813, 2736, 2551, 2541, 2834, 2260, 2366, 2275],
        'PRD_2': [1383, 1235, 1301, 1397, 1478, 1316, 1345, 1475, 1333, 1541, 1293, 1307, 1432, 1363, 1312, 918, 1219, 1432, 1413, 1286, 1358, 1552, 1243, 1320, 1122],
        'PRD_3': [4473, 4189, 4218, 4282, 4403, 4040, 4176, 4207, 4150, 4423, 3778, 3743, 4098, 3854, 3461, 2453, 3497, 4113, 3828, 3515, 3749, 3883, 3392, 3369, 3621],
        'Date': ['2019-01', '2019-02', '2019-03', '2019-04', '2019-05', '2019-06', '2019-07', '2019-08', '2019-09', '2019-10', '2019-11', '2019-12', '2020-01', '2020-02', '2020-03', '2020-04', '2020-05', '2020-06', '2020-07', '2020-08', '2020-09', '2020-10', '2020-11', '2020-12', '2021-01']
}

df = pd.DataFrame(data)

df['Date'] = pd.to_datetime(df['Date'])
df = df.set_index('Date')
df.index.freq = 'MS'


end_train = '2019-12'
start_test = '2020-01'
data_train = df.loc[:end_train, :].copy()
data_test = df.loc[start_test:, :].copy()

print(f"Frequency: {df.index.freq}")

forecaster = ForecasterAutoregMultiVariate(
    regressor = Ridge(random_state=123),
    level = 'PRD_1',
    lags = 3,
    steps = 6,
    transformer_series = None,
    transformer_exog = None,
    weight_func = None,
    n_jobs = 'auto'
)

forecaster.fit(series=data_train)
predictions = forecaster.predict()
print(predictions)
```

```
Frequency: <MonthBegin>
                  PRD_1
2020-01-01  3593.178420
2020-02-01  2750.850213
2020-03-01  2155.525411
2020-04-01  3235.457829
2020-05-01  2516.244082
2020-06-01  2897.071644`
```

**Reply by ABJ66** [reply]
Thanks for the response, Joaquin, your above solution works as expected.

However, I do like to mention the following steps I tried in the meantime: 
1. Read data as pandas df with index_col = ['Date'], 
2. Set index to Y-m using: df.index = pd.PeriodIndex(df.index, freq = 'M'),
3. Split df into data_train and data_test using relevant train_end and test_start Y-m as string,
4. On printing the data_train.index.freq, data_test.index.freq, I get this: "MonthEnd", "MonthEnd"
5. But it seems this does not work as the predictions still return integer based values for the indices

**Reply by JoaquinAmatRodrigo** [reply, marked as answer]
I see now. The problem is that the index is `PeriodIndex`  instead of `DatetimeIndex` (an index type I was not aware of). The current implementation of skforecast overwrites with a  `RangeIndex` everything that is not  `DatetimeIndex`. So nothing wrong from your side.
What I would suggest is to replace step 2 with:

```
df['Date'] = pd.to_datetime(df['Date'])
df = df.set_index('Date')
df.index.freq = 'MS'
```

Hope this help

**Comment by JavierEscobarOrtiz** [comment]
Hello @ABJ66,

In addition to Joaquín's answer, I suggest trying a `ForecasterAutoregMultiSeries`. This forecaster creates a single model that is able to predict all series at the same time.

https://skforecast.org/latest/user_guides/independent-multi-time-series-forecasting

**Reply by ABJ66** [reply]
Hi Javier, 

I will surely check this method as well. Thank you.
